coroutines/src/commonTest/kotlin/DelayStrategyTest.kt

```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Test
+    fun firstIteration_hasDelayEqualToBaseTimeMillis() {
+        assertEquals(100, getExponentialBackoffMillis(baseTimeMillis = 100, iteration = 0))
+    }
+
+    @Test
+    fun multipleRetries_increaseExponentially() {
+        val baseTimeMillis = 100L
+        var delay = baseTimeMillis
+        val iterations = 0..10
+        iterations.forEach { iteration ->
+            assertEquals(delay, getExponentialBackoffMillis(baseTimeMillis = 100L, iteration = iteration), "Backoff retry #$iteration")
+            delay *= 2
+        }
+    }
+
+    @Test
+    @OptIn(ExperimentalTime::class)
+    fun dynamicDelayStrategy_switchesBetweenStrategies_usingTrigger() = runTest {
+        val trigger = MutableStateFlow(true)
+        val timeMachine = TimeMachine()
+        val delayA = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 1_000L)
+        val delayB = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 2_000L)
+        val dynamic = Dynamic(trigger, timeMachine) { if (it) delayA else delayB }
+        val job = launch {
+            dynamic.await(iteration = 0, elapsedMillis = 0L)
+        }
+        assertEquals(
+            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 0L),
+            actual = delayA.invocation.receive(),
+        )
+        assertFalse(job.isCompleted)
+        timeMachine.advanceBy(milliseconds = 500L)
+        trigger.value = false
+        assertEquals(
+            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 500L),
+            actual = delayB.invocation.receive(),
+        )
+        assertFalse(job.isCompleted)
+        timeMachine.advanceBy(milliseconds = 3_000L)
+    }
+
+    @OptIn(ExperimentalTime::class)
+    private class TimeMachine : TimeSource {
+        private val elapsedMillis = MutableStateFlow(0L)
+        fun advanceBy(milliseconds: Long) { elapsedMillis.update { value -> value + milliseconds } }
+        suspend fun delayUntil(milliseconds: Long) { elapsedMillis.first { it >= milliseconds } }
+        private class TimeMachineMark(private val parent: TimeMachine) : TimeMark() {
+            override fun elapsedNow(): Duration = parent.elapsedMillis.value.toDuration(DurationUnit.MILLISECONDS)
+        }
+        override fun markNow(): TimeMark = TimeMachineMark(this)
+    }
+    private sealed class Invocation {
+        data class Await(val iteration: Int, val elapsedMilliseconds: Long) : Invocation()
+    }
+    private class TimeMachineDelayStrategy(
+        private val timeMachine: TimeMachine,
+        private val delayUntilMilliseconds: Long,
+    ) : DelayStrategy {
+        val invocation = Channel<Invocation>()
+        override suspend fun await(iteration: Int, elapsedMillis: Long) {
+            invocation.send(Invocation.Await(iteration, elapsedMillis))
+            timeMachine.delayUntil(milliseconds = delayUntilMilliseconds)
+        }
+    }
```

Comment:
There should be moved outside the test class, IMHO.
i.e. to the end of the file.